### PR TITLE
Issue 3666 context out of xhr

### PR DESCRIPTION
**Pull request by shane-tomlinson**
@6a68 - yo dude, I am trying to pull the context stuff out of xhr and move it into network.js. It has gone well, but I could use some collaboration time.

Specifically, https://github.com/shane-tomlinson/browserid/compare/issue_3666_context_out_of_xhr?expand=1#L1R53.  I want to change this to:

`get({`

but doing so has repercussions that I did not foresee. The old code requested session_context and returned results would be returned immediately instead of being deferred. This is the _only_ request that acted like this. I want to make this request consistent with all other requests and have its completion deferred. 

Some tests depend on this behavior, the tests are in `user.js` and `state.js`. You can see in the first commit the changes I had to make to the tests. I was a bit uncomfortable doing this without some input to see if the scope of the PR is expanding too far.

This is a continuation of mozilla/browserid#3655 to address mozilla/browserid#3666.

### CODE REVIEW

File: resources/static/common/js/network.js [hunk 1]
```diff
@@ -8,57 +8,70 @@ BrowserID.Network = (function() {
   var bid = BrowserID,
       complete = bid.Helpers.complete,
       context,
-      server_time,
```
Comment (shane-tomlinson) [inline resources/static/common/js/network.js:11]: These three fields are also stored in the context variable. too many sources of truth hurt my brain, so I removed the duplicates.

File: resources/static/common/js/network.js [hunk 2]
```diff
@@ -8,57 +8,70 @@ BrowserID.Network = (function() {
...
       XHR = bid.Modules.XHR,
       xhr,
       // XXX get this out of here!
       storage = bid.Storage;
 
-  function post() {
-    xhr.post.apply(xhr, [].slice.call(arguments, 0));
-  }
+  function post(options) {
+    if (!context) {
```
Comment (shane-tomlinson) [inline resources/static/common/js/network.js:19]: Do the context check here in network.js instead of xhr.js, this makes more sense to me.

File: resources/static/common/js/network.js [hunk 3]
```diff
@@ -8,57 +8,70 @@ BrowserID.Network = (function() {
...
+      document.cookie = "can_set_cookies=1";
+    } catch(e) {
+      // If cookies are disabled, some browsers throw an exception. Ignore
+      // this, the backend will see that cookies are disabled.
     }
+
+    // use the low level request because get defers success which jacks up
+    // a bunch of tests that expect this to return immediately. Who coded this?
+    // (me, I shrug embarrassed).
+    xhr.request({
```
Comment (shane-tomlinson) [inline resources/static/common/js/network.js]: This is the line in question, I want to change this to be just `get`, but tests erroneously expect this call to immediately return when using a mock low level XHR object.

File: resources/static/common/js/network.js [hunk 4]
```diff
@@ -8,57 +8,70 @@ BrowserID.Network = (function() {
...
+      local_time: new Date().getTime()
+    });
+
+    mediator.publish("context_info", newContext);
   }
 
   function clearContext() {
-    if (xhr) xhr.clearContext();
     var undef;
-    context = server_time = undef;
```
Comment (shane-tomlinson) [inline resources/static/common/js/network.js:61]: We were incorrectly not clearing the multiple sources of truth, and that hurt my head and prompted me to only use context.

File: resources/static/common/js/network.js [hunk 5]
```diff
@@ -168,7 +179,13 @@ BrowserID.Network = (function() {
     withContext: withContext,
 
     setContext: function(field, value) {
```
Comment (shane-tomlinson) [inline resources/static/common/js/network.js:178]: setContext is used in testing, when used in testing, the entire context object can be set at once.

File: resources/static/test/cases/common/js/network.js [hunk 6]
```diff
@@ -553,18 +553,22 @@
   asyncTest("cookiesEnabled with onComplete exception thrown - should not call onComplete a second time", function() {
     // Since we are manually throwing an exception, it must be caught
     // below.
-    try {
-      network.cookiesEnabled(function(status) {
-        // if there is a problem, this callback will be called a second time
-        // with a false status.
-        equal(status, true, "cookies are enabled, correct status");
-        start();
+    network.withContext(function() {
```
Comment (shane-tomlinson) [inline resources/static/test/cases/common/js/network.js:556]: This test was just plain out not testing what it needed to test. You can't check for whether an exception was thrown from within the catch block, it might never be called.

File: resources/static/common/js/network.js [hunk 7]
```diff
@@ -8,57 +8,70 @@ BrowserID.Network = (function() {
...
+      error: onFailure
+    });
+  }
+
+  function setContext(newContext) {
+    context = _.extend({}, newContext, {
+      local_time: new Date().getTime()
+    });
+
+    mediator.publish("context_info", newContext);
```
Comment (jaredhirsch) [inline resources/static/common/js/network.js]: should mediator publish context here, rather than newContext? do we want local_time in the published signal?

File: resources/static/common/js/network.js (hunk 5 above)
Comment (jaredhirsch) [inline resources/static/common/js/network.js:178]: I'm trying to imagine if this could accidentally bite us, say if you didn't pass a second argument. Is it possible to check if we're in test mode before allowing this? /me squirms

File: resources/static/test/cases/common/js/network.js (hunk 6 above)
Comment (jaredhirsch) [inline resources/static/test/cases/common/js/network.js:556]: yup, good catch

File: resources/static/common/js/network.js (hunk 5 above)
Comment (shane-tomlinson) [inline resources/static/common/js/network.js:178]: mmm.. What if instead of surfacing a testing variable, if we check if arguments.length === 1 and if the first argument is an object?

File: resources/static/common/js/modules/xhr.js [hunk 8]
```diff
@@ -147,9 +135,11 @@ BrowserID.Modules.XHR = (function() {
      * the user informing them of the slowness.
      */
     if (self.time_until_delay) {
-      request.slowRequestTimeout = setTimeout(function() {
+      var timer = request.slowRequestTimeout = setTimeout(function() {
```
Comment (shane-tomlinson) [inline resources/static/common/js/modules/xhr.js:150]: All setTimeout timers are kept track of and aborted when the module stops to prevent cross-test contamination. This became a problem when the call to /wsapi/session_context started using "defer_success" like all the other requests.

File: resources/static/common/js/network.js [hunk 9]
```diff
@@ -6,59 +6,69 @@ BrowserID.Network = (function() {
...
-      // to ensure the cookie is set for that single call.
-      try {
-        document.cookie = "can_set_cookies=1";
-      } catch(e) {
-        // If cookies are disabled, some browsers throw an exception. Ignore
-        // this, the backend will see that cookies are disabled.
-      }
-      xhr.getContext(cb, onFailure);
+  function withContext(done, onFailure) {
+    if (typeof context !== "undefined") return complete(done, context);
```
Comment (jaredhirsch) [inline resources/static/common/js/network.js]: inside `post` we do the falsy check, `if (!context)`. Frankly, I prefer a looser check unless you explicitly want to distinguish between falsy types, but our jshint settings seem to imply that we prefer super-strict checks under all circumstances. So either let's loosen this one or tighten the check in `post`.

File: resources/static/common/js/network.js [hunk 10]
```diff
@@ -6,59 +6,69 @@ BrowserID.Network = (function() {
...
+    });
+
+    mediator.publish("context_info", context);
   }
 
   function clearContext() {
-    if (xhr) xhr.clearContext();
     var undef;
-    context = server_time = undef;
+    context = undef;
```
Comment (jaredhirsch) [inline resources/static/common/js/network.js:71]: I like this undef thing (unset things by assigning to an uninitialized variable instead of 'undefined' itself), but it's at odds with straight-up setting to `undefined` elsewhere in this PR. Probably we should be consistent about how we unset things--I prefer this approach because it's more minifiable, but your call on which to use.

File: resources/static/common/js/network.js [hunk 11]
```diff
@@ -168,7 +176,15 @@ BrowserID.Network = (function() {
...
     setContext: function(field, value) {
-      if (context) context[field] = value;
+      if (arguments.length === 1) {
+        if (!helpers.isObject(field)) throw new Error("invalid context");
+
+        // an object was passed in for the context. Used for testing.
+        setContext(field);
+      }
+      else {
+        if (context) context[field] = value;
```
Comment (jaredhirsch) [inline resources/static/common/js/network.js:186]: again--here's a truthy check. if we want to be uniformly strict, I suppose `isObject(context)` or `typeof context !== 'undefined'` should be used instead.

File: resources/static/common/js/network.js [hunk 12]
```diff
@@ -579,13 +595,14 @@ BrowserID.Network = (function() {
...
+        var server_time = context.server_time;
         try {
           if (!server_time) throw new Error("can't get server time!");
-          var offset = (new Date()).getTime() - server_time.local;
-          complete(onComplete, new Date(offset + server_time.remote));
+          var offset = (new Date()).getTime() - context.local_time;
+          complete(onComplete, new Date(offset + server_time));
         } catch(e) {
-          complete(onFailure, e.toString());
+          complete(onFailure, String(e));
```
Comment (jaredhirsch) [inline resources/static/common/js/network.js:605]: I'm not sure I follow this change. Why's this a better way? I'm not familiar with it.

File: resources/static/common/js/network.js (hunk 5 above)
Comment (jaredhirsch) [inline resources/static/common/js/network.js:178]: sure, let's do it.

File: resources/static/common/js/network.js [hunk 13]
```diff
@@ -600,12 +617,13 @@ BrowserID.Network = (function() {
...
     domainKeyCreationTime: function(onComplete, onFailure) {
-      withContext(function() {
+      withContext(function(context) {
+        var domain_key_creation_time = context.domain_key_creation_time;
         try {
           if (!domain_key_creation_time) throw new Error("can't get domain key creation time!");
           complete(onComplete, new Date(domain_key_creation_time));
         } catch(e) {
-          complete(onFailure, e.toString());
+          complete(onFailure, String(e));
```
Comment (jaredhirsch) [inline resources/static/common/js/network.js:626]: ditto

File: resources/static/dialog/js/misc/state.js [hunk 14]
```diff
@@ -261,6 +261,7 @@ BrowserID.State = (function() {
         startAction(false, "doStageUser", info);
 // TODO         startAction(false, "doStageResetPassword", info); ???
```
Comment (jaredhirsch) [inline resources/static/dialog/js/misc/state.js]: too much scope creep to address this TODO while we're here? your call, I'm not sure.

File: resources/static/dialog/js/misc/state.js [hunk 15]
```diff
@@ -261,6 +261,7 @@ BrowserID.State = (function() {
...
       }
+      complete(info.complete);
```
Comment (jaredhirsch) [inline resources/static/dialog/js/misc/state.js:263]: This seems significant. Can you explain a bit why this has been added?

File: resources/static/test/cases/common/js/helpers.js [hunk 16]
```diff
@@ -162,4 +162,35 @@
...
+    equal(helpers.isFunction([]), false);
+  });
+
+  test("isArray", function() {
+    var someFunc = function() {};
+    equal(helpers.isArray([]), true);
+    equal(helpers.isArray(new Array()), true);
+    equal(helpers.isArray({}), false);
+    equal(helpers.isArray(""), false);
+  });
```
Comment (jaredhirsch) [inline resources/static/test/cases/common/js/helpers.js:196]: one of the beauties of our `isArray` implementation is that it correctly distinguishes arrays from non-Array things with length, like a DOM NodeList or an `arguments` object. Maybe add tests for those here? `arguments` should be easy, the DOM bit might be more trouble than it's worth.

_edit: derp._

File: resources/static/test/cases/dialog/js/misc/state.js [hunk 17]
```diff
@@ -55,12 +55,19 @@
...
+        testActionStarted("doSetPassword");
+        mediator.publish("password_set", {
+          complete: function() {
+            equal(actions.info[stagingAction].email, TEST_EMAIL,
+                      "correct email sent to " + stagingAction);
+            start();
+          }
+        });
+      }
+    });
```
Comment (jaredhirsch) [inline resources/static/test/cases/dialog/js/misc/state.js:70]: hmm, this seems connected to the mysterious `complete` call inside `state.js`. Is this another thing that accidentally got pushed, or just something you've fixed while you were in there?

File: resources/static/common/js/modules/xhr.js [hunk 18]
```diff
@@ -169,59 +159,50 @@ BrowserID.Modules.XHR = (function() {
 
   function get(options) {
     /*jshint validthis: true*/
-    var req = _.extend(options, {
+    options = _.extend(options, {
       type: "GET",
       defer_success: true
     });
-    this.request(req);
+    this.request(options);
```
Comment (jaredhirsch) [inline resources/static/common/js/modules/xhr.js]: `makeRequest` returns the request object, but `get` and `post` swallow it and return undefined.

Would it be useful to return the request upward to callers? Or is it a leaky abstraction asking for trouble? I'm not sure. Either way, we probably should be consistent, as `makeRequest` is part of the exported module API along with `get` and `post`. Your call.

Actually, it looks like we publish the request in the `xhr_start`, `xhr_complete`, and `xhr_delay` signals, so I think `get` and `post` should return `this.request()`? Yay uniformity.

File: resources/static/common/js/modules/xhr.js (hunk 8 above)
Comment (jaredhirsch) [inline resources/static/common/js/modules/xhr.js:150]: <3

**Comment by jaredhirsch** [comment]
@shane-tomlinson Welp, I've had quite a nit-picking fiesta at your expense. I'm kinda fried at the moment, can't provide good high-level input. But xhr.js is looking really sleek at this point, :beers: all around.

Tests passing for me locally, and Travis seems happy too.

R+ modulo assorted minor complaints. Feel free to merge in after addressing (or dismissing, as the case may be)

**Comment by shane-tomlinson** [comment]
Thanks for the awesome review @6a68, time for some updates!

File: resources/static/common/js/network.js (hunk 12 above)
Comment (shane-tomlinson) [inline resources/static/common/js/network.js:605]: @seanmonstar convinced me that String(e) was the way to go when printing exceptions for the reasons outlined in http://stackoverflow.com/questions/3945202/whats-the-difference-between-stringvalue-vs-value-tostring - mainly for when we start passing exceptions around. Even though we are guaranteed to have `e` here, I changed these to be consistent with what we have done in other areas.

File: resources/static/common/js/modules/xhr.js (hunk 18 above)
Comment (shane-tomlinson) [inline resources/static/common/js/modules/xhr.js]: Good point. I originally did not return the XHR object because `post` was taking care of fetching the CSRF token, which resulted in 2 requests. That is now taken care of in network, meaning `post` can safely return the object.

File: resources/static/common/js/network.js (hunk 9 above)
Comment (shane-tomlinson) [inline resources/static/common/js/network.js]: Updating & testing.

File: resources/static/dialog/js/misc/state.js (hunk 14 above)
Comment (shane-tomlinson) [inline resources/static/dialog/js/misc/state.js]: I'm not sure _what_ that line is supposed to do. I think it can now safely be removed, as we aren't seeing any issues with new FFOS accounts.

File: resources/static/test/cases/dialog/js/misc/state.js (hunk 17 above)
Comment (shane-tomlinson) [inline resources/static/test/cases/dialog/js/misc/state.js:70]: Dude, you made me dig deep to figure out why I made this change.

Here's what happened:

The mocked call to /wsapi/session_context used to return synchronously, we made the change to return asynchronously like all the other calls. This caused this test, which previously depended on the synchronous behavior in the `mediator.publish(stagingMessage..` to return immediately. This is now no longer the case, and and our assumption was broken. So, now, depend on the asynchronous behavior and the fact that the complete message on R60 will be called when the staging has completed. The password_set bit was changed to asynchronous behavior just to keep consistency, it is not strictly necessary, but I like consistent behavior so I don't have to remember special cases.

Phew. I can relax again.

File: resources/static/test/cases/common/js/helpers.js (hunk 16 above)
Comment (shane-tomlinson) [inline resources/static/test/cases/common/js/helpers.js:196]: Added a test for arguments.
